File: 카카오/20인턴십/키패드누르기.py
numbers = list(map(int,input()))
hand = input()

def solution(numbers, hand):
    answer = ''
    arr = [[0 for col in range(3)] for row in range(4)]
    left = [1,4,7]
    left_pos = [(0,0),(1,0),(2,0)]
    right = [3,6,9]
    right_pos = [(0,2),(1,2),(2,2)]
    middle = [2,5,8,0]
    middle_pos = [(0,1),(1,1),(2,1),(3,1)]

    point = [[3,0],[3,2]]
    for i in range(len(numbers)):
        if numbers[i] in left:
            point[0] = left_pos[left.index(numbers[i])]
            answer+='L'
        elif numbers[i] in right:
            point[1] = right_pos[right.index(numbers[i])]
            answer+='R'
        elif numbers[i] in middle:
            o_point = middle_pos[middle.index(numbers[i])]
            left_distance = abs(point[0][0]-o_point[0]) + abs(point[0][1]-o_point[1])
            right_distance = abs(point[1][0]-o_point[0]) + abs(point[1][1]-o_point[1])
            if(left_distance<right_distance):
                point[0] = o_point
                answer+='L'
            elif(left_distance>right_distance):
                point[1] = o_point
                answer+='R'
            elif(left_distance == right_distance):
                if(hand == 'left'):
                    point[0] = o_point
                    answer+='L'
                else:
                    point[1] = o_point
                    answer+='R'
    
    return answer

result = solution(numbers, hand)
print(result)

File: 프로그래머스/Level1/키패드누르기.java
import java.io.*;
import java.util.*;

class Solution {
    public String solution(int[] numbers, String hand) {
        String answer = "";
        Hand left = new Hand(3,0);
        Hand right = new Hand(3,2);
        
        StringBuilder sb = new StringBuilder();
        for(int number : numbers) {
        	if(number==0)
        		number=11;
        	
        	if(number==1 || number==4 || number==7) {
        		sb.append('L');
        		left.x = (int)((number-1)/3);
        		left.y = (number-1)%3;
        	}
        	else if(number==3 || number==6 || number==9) {
        		sb.append('R');
        		right.x = (int)((number-1)/3);
        		right.y = (number-1)%3;
        	}
        	else {
        		Hand tmp = new Hand(number/3, 1);
        		
        		int leftDistance=Math.abs(left.x-tmp.x) + Math.abs(left.y-tmp.y);
        		int rightDistance=Math.abs(right.x-tmp.x) + Math.abs(right.y-tmp.y);
        		if(leftDistance<rightDistance)
        		{
        			left.x=tmp.x; left.y=tmp.y;
        			sb.append('L');
        		}
        		else if(leftDistance>rightDistance) {
        			right.x=tmp.x; right.y = tmp.y;
        			sb.append('R');
        		}
        		else {
        			if(hand.equals("left")) {
        				sb.append('L');
        				left.x=tmp.x; left.y=tmp.y;
        			}
        			else {
        				sb.append('R');
        				right.x=tmp.x; right.y = tmp.y;
        			}
        		}
        	}
        	
        }
        
        return sb.toString();
    }
    
    static class Hand{
    	int x,y;
    	Hand(int x, int y){
    		this.x=x;
    		this.y=y;
    	}
    }
}

public class Main {
	public static void main(String[] args) throws IOException {
		// TODO Auto-generated method stub
		int[] numbers = {7, 0, 8, 2, 8, 3, 1, 5, 7, 6, 2};
		String hand = "left";
		System.out.println(new Solution().solution(numbers,hand));		
	}

}
